transport/src/main/java/io/netty/channel/PendingWriteQueue.java

```diff
@@ -37,22 +38,69 @@
             SystemPropertyUtil.getInt("io.netty.transport.pendingWriteSizeOverhead", 64);
 
     private final ChannelHandlerContext ctx;
-    private final ChannelOutboundBuffer buffer;
-    private final MessageSizeEstimator.Handle estimatorHandle;
+    private final PendingTracker tracker;
 
     // head and tail pointers for the linked-list structure. If empty head and tail are null.
     private PendingWrite head;
     private PendingWrite tail;
     private int size;
     private long bytes;
 
+    private interface PendingTracker extends MessageSizeEstimator.Handle {
+        void incrementPendingOutboundBytes(long bytes);
+        void decrementPendingOutboundBytes(long bytes);
+    }
+
     public PendingWriteQueue(ChannelHandlerContext ctx) {
-        if (ctx == null) {
-            throw new NullPointerException("ctx");
+        this.ctx = ObjectUtil.checkNotNull(ctx, "ctx");
+        if (ctx.pipeline() instanceof DefaultChannelPipeline) {
```

Comment:
please note I am not super happy with this but I think its ok for now. But we may need to find a more proper way at some point :)